### PR TITLE
Retry Infinispan factory creation on timeouts

### LINKED ISSUE #16671: Flaky test: org.keycloak.testsuite.model.session.OfflineSessionPersistenceTest#testPersistenceMultipleNodesClientSessionsAtRandomNode
## org.keycloak.testsuite.model.session.OfflineSessionPersistenceTest#testPersistenceMultipleNodesClientSessionsAtRandomNode
[Keycloak CI - Store Model Tests](https://github.com/keycloak/keycloak/actions/runs/3994036682/jobs/6851401076) / [Pull Request #16478](https://github.com/keycloak/keycloak/pull/16478)

```
java.lang.AssertionError: 
threads didn't terminate in time: [main (TIMED_WAITING): locked on java.util.concurrent.FutureTask@77fc5ab7 waiting for 96147 ms blocked for 0 ms
	at java.base@11.0.17/jdk.internal.misc.Unsafe.park(Native Method)
	at java.base@11.0.17/java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:234)
	at java.base@11.0.17/java.util.concurrent.FutureTask.awaitDone(FutureTask.java:444)
...
```

### PR DESCRIPTION
**Pull request by ahus1**
Relates to #16671

### CODE REVIEW

**Comment by ahus1** [comment]
I've seen one of my GitHub action runs fail with this exception. Afterwards the tests didn't recover with a retry. The action failed with the "red" status, and therefore also failed the required check. 

https://github.com/keycloak/keycloak/actions/runs/4357050101/jobs/7616139970

This PR adds:
* Retry procedure for the tests when the startup fails.
* Cleanup of a failed setup of a factory in the tests.

This change aims to make no changes to the production code, only to the test code. Such a failure could still happen during a regular start of Keycloak: In such a case the start wouldn't complete. On a setup like Kubernetes, the Pod would be restarted, so I assume this is ok in such a scenario. It is probably not ok in other scenarios, those would need to add a retry-method to the Keycloak wrapper.

```
2023-03-07T18:27:44.0004725Z 18:27:43,977 DEBUG [org.keycloak.connections.jpa.DefaultJpaConnectionProviderFactory] (pool-11-thread-3) Calling migrateModel
2023-03-07T18:27:44.0006158Z 18:27:43,979 ERROR [org.keycloak.testsuite.model.KeycloakModelParameters] (Time-limited test) Execution didn't complete
2023-03-07T18:27:44.0007510Z java.util.concurrent.ExecutionException: java.lang.RuntimeException: org.infinispan.util.concurrent.TimeoutException: ISPN000658: Initial state transfer timed out for cache offlineSessions on node-9
2023-03-07T18:27:44.0008584Z    at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
2023-03-07T18:27:44.0009643Z    at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:205)
2023-03-07T18:27:44.0010560Z    at org.keycloak.testsuite.model.KeycloakModelTest.inIndependentFactories(KeycloakModelTest.java:427)
2023-03-07T18:27:44.0075816Z    at org.keycloak.testsuite.model.session.OfflineSessionPersistenceTest.testPersistenceMultipleNodesClientSessionsAtRandomNode(OfflineSessionPersistenceTest.java:246)
2023-03-07T18:27:44.0098029Z    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2023-03-07T18:27:44.0098704Z    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
2023-03-07T18:27:44.0099416Z    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2023-03-07T18:27:44.0100089Z    at java.base/java.lang.reflect.Method.invoke(Method.java:568)
2023-03-07T18:27:44.0100614Z    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
2023-03-07T18:27:44.0101228Z    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
2023-03-07T18:27:44.0101855Z    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
2023-03-07T18:27:44.0102530Z    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
2023-03-07T18:27:44.0103194Z    at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
2023-03-07T18:27:44.0103850Z    at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
2023-03-07T18:27:44.0104414Z    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2023-03-07T18:27:44.0104840Z    at java.base/java.lang.Thread.run(Thread.java:833)
2023-03-07T18:27:44.0105763Z Caused by: java.lang.RuntimeException: org.infinispan.util.concurrent.TimeoutException: ISPN000658: Initial state transfer timed out for cache offlineSessions on node-9
2023-03-07T18:27:44.0106540Z    at org.keycloak.testsuite.model.KeycloakModelTest.inIndependentFactory(KeycloakModelTest.java:498)
2023-03-07T18:27:44.0107257Z    at org.keycloak.testsuite.model.KeycloakModelTest.lambda$inIndependentFactories$6(KeycloakModelTest.java:398)
2023-03-07T18:27:44.0107845Z    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2023-03-07T18:27:44.0108400Z    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
2023-03-07T18:27:44.0109024Z    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
2023-03-07T18:27:44.0109432Z    ... 1 more
2023-03-07T18:27:44.0110013Z Caused by: org.infinispan.util.concurrent.TimeoutException: ISPN000658: Initial state transfer timed out for cache offlineSessions on node-9
2023-03-07T18:27:44.0111333Z    at org.infinispan.statetransfer.StateTransferManagerImpl.waitForInitialStateTransferToComplete(StateTransferManagerImpl.java:239)
2023-03-07T18:27:44.0112113Z    at org.infinispan.cache.impl.CacheImpl.start(CacheImpl.java:1013)
2023-03-07T18:27:44.0129129Z    at org.infinispan.cache.impl.AbstractDelegatingCache.start(AbstractDelegatingCache.java:504)
2023-03-07T18:27:44.0129849Z    at org.infinispan.manager.DefaultCacheManager.wireAndStartCache(DefaultCacheManager.java:722)
2023-03-07T18:27:44.0130541Z    at org.infinispan.manager.DefaultCacheManager.createCache(DefaultCacheManager.java:668)
2023-03-07T18:27:44.0131223Z    at org.infinispan.manager.DefaultCacheManager.internalGetCache(DefaultCacheManager.java:557)
2023-03-07T18:27:44.0131889Z    at org.infinispan.manager.DefaultCacheManager.getCache(DefaultCacheManager.java:571)
2023-03-07T18:27:44.0132798Z    at org.keycloak.connections.infinispan.DefaultInfinispanConnectionProviderFactory.initEmbedded(DefaultInfinispanConnectionProviderFactory.java:315)
2023-03-07T18:27:44.0134187Z    at org.keycloak.connections.infinispan.DefaultInfinispanConnectionProviderFactory.lazyInit(DefaultInfinispanConnectionProviderFactory.java:150)
2023-03-07T18:27:44.0135303Z    at org.keycloak.connections.infinispan.DefaultInfinispanConnectionProviderFactory.create(DefaultInfinispanConnectionProviderFactory.java:83)
2023-03-07T18:27:44.0136652Z    at org.keycloak.connections.infinispan.DefaultInfinispanConnectionProviderFactory.create(DefaultInfinispanConnectionProviderFactory.java:67)
2023-03-07T18:27:44.0137535Z    at org.keycloak.services.DefaultKeycloakSession.getProvider(DefaultKeycloakSession.java:271)
2023-03-07T18:27:44.0138536Z    at org.keycloak.models.sessions.infinispan.InfinispanSingleUseObjectProviderFactory.getSingleUseObjectCache(InfinispanSingleUseObjectProviderFactory.java:53)
2023-03-07T18:27:44.0139743Z    at org.keycloak.models.sessions.infinispan.InfinispanSingleUseObjectProviderFactory.postInit(InfinispanSingleUseObjectProviderFactory.java:77)
2023-03-07T18:27:44.0140664Z    at org.keycloak.services.DefaultKeycloakSessionFactory.init(DefaultKeycloakSessionFactory.java:128)
2023-03-07T18:27:44.0141325Z    at org.keycloak.testsuite.model.KeycloakModelTest$4.init(KeycloakModelTest.java:322)
2023-03-07T18:27:44.0142035Z    at org.keycloak.testsuite.model.KeycloakModelTest.createKeycloakSessionFactory(KeycloakModelTest.java:349)
2023-03-07T18:27:44.0142888Z    at org.keycloak.testsuite.model.KeycloakModelTest.reinitializeKeycloakSessionFactory(KeycloakModelTest.java:359)
2023-03-07T18:27:44.0143943Z    at org.keycloak.testsuite.model.session.OfflineSessionPersistenceTest.lambda$testPersistenceMultipleNodesClientSessionsAtRandomNode$27(OfflineSessionPersistenceTest.java:274)
2023-03-07T18:27:44.0144894Z    at org.keycloak.testsuite.model.KeycloakModelTest.lambda$inIndependentFactories$5(KeycloakModelTest.java:405)
2023-03-07T18:27:44.0145636Z    at org.keycloak.testsuite.model.KeycloakModelTest.inIndependentFactory(KeycloakModelTest.java:496)
2023-03-07T18:27:44.0146125Z    ... 5 more
```

**Comment by ahus1** [comment]
@mhajas / @martin-kanis: One of my previous GHA was unstable, and I looked into the logs why this was the case. I'd like to make those changes. 

While the build is now green, it is difficult to see if this gets it more stable - I looked at the logs and the new log message didn't appear. I'd still like to make those changes and then see how this behaves. 

I'd be happy if one of you could review this. Please add yourself as a reviewer, as I don't know who of you would be interested or have the time to review it. Thanks!

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/KeycloakModelTest.java [hunk 1]
```diff
@@ -302,8 +302,11 @@ protected void finished(Description description) {
      * local to the thread that calls this method, allowing for per-thread customization. This in turn allows
      * testing of several parallel session factories which can be used to simulate several servers
      * running in parallel.
+     * <p>
+     * The method itself is synchronized as <code>CONFIG</code> and <code>MODEL_PARAMETERS</code> are shared resources
+     * used during the initialization of the factories.
      */
-    public static KeycloakSessionFactory createKeycloakSessionFactory() {
+    public static synchronized KeycloakSessionFactory createKeycloakSessionFactory() {
```
Comment (martin-kanis) [inline testsuite/model/src/test/java/org/keycloak/testsuite/model/KeycloakModelTest.java:309]: Does this effectively mean that `inIndependentFactories()` will now spawn new nodes sequentially? And if so, do we want it?

Comment (ahus1) [reply testsuite/model/src/test/java/org/keycloak/testsuite/model/KeycloakModelTest.java]: Yes, @martin-kanis that was my intention when I created this PR as I thought CONFIG wouldn't be thread safe. I now looked at it again, and saw that will use a thread safe implementation when it is not running on the main thread. Therefore I reverted that part, and only keep the retry login the updated PR. Please let me know if this now looks better to you and can be approved.

Comment (martin-kanis) [reply testsuite/model/src/test/java/org/keycloak/testsuite/model/KeycloakModelTest.java]: Thanks for the explanation and changes. I scheduled [a run in GHA](https://github.com/martin-kanis/keycloak/actions/runs/4415190525/jobs/7737848284) with 10 model test executions to see if we get any retry triggered.

**Review by martin-kanis — DISMISSED** [review]

**Review by mhajas — COMMENTED** [review]
@ahus1 thank you for the PR! However, is this the correct thing to do? Isn't this just hiding some real instability issue by adding more retries? 

If I remember correctly when this test used to time out so we made it fail as part of this #16427 so we can monitor what to do next. I see that this test failed ~18 times in two months.

I think we should proceed with this fix only if we are sure this happens only in model tests, if not, we should probably investigate it properly and maybe fix it for KC22 if needed. WDYT? 

cc: @martin-kanis @hmlnarik

**Comment by ahus1** [comment]
@martin-kanis / @mhajas - the primary reason I opened this PR was that a GitHub action for the model tests failed completely (and not even retried worked). 

If we would choose to not have a retry added as part of this PR, I'd argue to have at least the cleanup after a failed Infinispan startup about closing the caches in here. Then I'd hope that the whole action will no longer fail. 

WDYT?

**Comment by ahus1** [comment]
@mhajas - sorry for replying so late. 

> I'd argue to have at least the cleanup after a failed Infinispan startup about closing the caches in here.

I've updated the PR to remove the retry. The change you now see is that after a failed session init, there is now a close that was previously missing. Without that, there would be additional Infinispan instances around which could negatively affect a retry / other tests in the suite. 

I hope this resolves some of the failure we see when retry doesn't work.

**Review by hmlnarik — APPROVED** [review]
The proposed change is correct, regardless of whether it has impact on the stability of infinispan tests. Thank you for that.

**Review by mhajas — APPROVED** [review]
Thank you @ahus1, this change makes a lot of sense

**Pull request merged**